File: ex10_c_dev_kit/examples/prbs_example.c
/*****************************************************************************
 *                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      *
 *                                                                           *
 * This source code is the property of Impinj, Inc. Your use of this source  *
 * code in whole or in part is subject to your applicable license terms      *
 * from Impinj.                                                              *
 * Contact support@impinj.com for a copy of the applicable Impinj license    *
 * terms.                                                                    *
 *                                                                           *
 * (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               *
 *                                                                           *
 *****************************************************************************/

#include <stdio.h>

#include "board/time_helpers.h"
#include "ex10_api/application_registers.h"
#include "ex10_api/board_init.h"
#include "ex10_api/rf_mode_definitions.h"
#include "ex10_api/trace.h"


/* Settings used when running this example */
const uint32_t prbs_time_on        = 8 * 1000;  // Duration in milliseconds
const uint8_t  antenna             = 1;
const uint16_t rf_mode             = mode_148;
const uint16_t transmit_power_cdbm = 3000;

int prbs_example(struct Ex10Interfaces ex10_iface)
{
    printf("Starting PRBS test\n");

    bool     transmitting = false;
    uint32_t start_time   = get_ex10_time_helpers()->time_now();
    while (get_ex10_time_helpers()->time_elapsed(start_time) < prbs_time_on)
    {
        if (!transmitting)
        {
            uint32_t const frequency_khz = 0u;  // frequency from hopping table
            uint32_t const remain_on     = false;  // Use regulatory times
            struct OpCompletionStatus op_status =
                ex10_iface.reader->prbs_test(antenna,
                                             rf_mode,
                                             transmit_power_cdbm,
                                             frequency_khz,
                                             remain_on);
            if (!ex10_iface.helpers->check_ops_status_errors(op_status))
            {
                return -1;
            }
            transmitting = true;
        }

        struct EventFifoPacket const* packet = ex10_iface.reader->packet_peek();
        if (packet)
        {
            if (packet->packet_type == TxRampDown)
            {
                transmitting = false;
            }
            ex10_iface.reader->packet_remove();
        }
    }

    ex10_iface.reader->stop_transmitting();
    printf("Ending PRBS test\n");
    return 0;
}

int main(void)
{
    tracepoint(pi_ex10sdk, EXEC_start, __FILE__);
    struct Ex10Interfaces ex10_iface =
        ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, "FCC");

    int result = ex10_iface.helpers->check_board_init_status(Application);
    if (result == 0)
    {
        result = prbs_example(ex10_iface);
    }

    ex10_typical_board_teardown();
    tracepoint(pi_ex10sdk, EXEC_end, __FILE__);

    return result;
}

File: ex10_dev_kit/examples/prbs_example.py
#############################################################################
#                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      #
#                                                                           #
# This source code is the property of Impinj, Inc. Your use of this source  #
# code in whole or in part is subject to your applicable license terms      #
# from Impinj.                                                              #
# Contact support@impinj.com for a copy of the applicable Impinj license    #
# terms.                                                                    #
#                                                                           #
# (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               #
#                                                                           #
#############################################################################
"""
Example code to use the PRBS functionality
"""

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)
from datetime import datetime

from py2c_interface.py2c_python_wrapper import *


PRBS_TIME_ON = 6                    # Duration of PRBS operation (seconds)
TRANSMIT_POWER_DBM = 3000           # 30 dBm
RF_MODE = RfModes.mode_5
REGION = 'FCC'                      # Regulatory region
R807_ANTENNA_PORT = 1               # Which R807 antenna port will be used
FREQUENCY_USED = 0                  # Set to 0 to use the jump table
REMAIN_ON = True


def run_prbs(ex10_ifaces):
    # pylint: disable=missing-docstring
    ex10_reader = ex10_ifaces.reader
    helper = ex10_ifaces.helpers

    print('Starting prbs test')

    try:
        transmitting = False
        start_time = datetime.now()
        time_delta = datetime.now() - start_time

        # Run PRBS
        while time_delta.total_seconds() <  PRBS_TIME_ON:
            if not transmitting:
                transmitting = True
                # Run PRBS with no regulatory timers to allow easier
                # viewing on a spectrum analyzer
                op_status = ex10_reader.prbs_test(
                    R807_ANTENNA_PORT,
                    RF_MODE,
                    TRANSMIT_POWER_DBM,
                    FREQUENCY_USED,
                    REMAIN_ON)
                assert op_status.error_occurred == False

            while ex10_reader.packets_available():
                packet = ex10_reader.packet_peek().contents
                helper.print_packets(packet)
                
                if packet.packet_type == EventPacketType.TxRampDown:
                    transmitting = False

                ex10_reader.packet_remove()

            time_delta = datetime.now() - start_time
    
    finally:
        ex10_reader.stop_transmitting()
        print('Ending prbs test')

def run_prbs_example():
    # pylint: disable=missing-docstring
    try:
        # Init the python to C layer
        py2c = Ex10Py2CWrapper()
        ex10_ifaces = py2c.ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, REGION.encode('ascii'))
        run_prbs(ex10_ifaces)
    finally:
        py2c.ex10_typical_board_teardown()


if __name__ == "__main__":
    run_prbs_example()
